File: excelium-sheets/src/main/java/excelium/sheets/connection/GoogleConnection.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Excelium
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package excelium.sheets.connection;

import com.google.api.client.auth.oauth2.Credential;
import com.google.api.client.googleapis.auth.oauth2.GoogleClientSecrets;
import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.services.sheets.v4.SheetsScopes;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;

/**
 * Represents connection to Google services.
 *
 * @author PhungDucKien
 * @since 2018.04.10
 */
public abstract class GoogleConnection {

    /**
     * Logger
     */
    private static final Logger LOG = LogManager.getLogger();

    /**
     * Global instance of the scopes.
     */
    protected static final List<String> SCOPES = Arrays.asList(SheetsScopes.SPREADSHEETS);

    /**
     * Global instance of the JSON factory.
     */
    public static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();

    /**
     * Global instance of the HTTP transport.
     */
    public static HttpTransport HTTP_TRANSPORT;

    /**
     * Construct a Google connection.
     */
    public GoogleConnection() {
        try {
            HTTP_TRANSPORT = GoogleNetHttpTransport.newTrustedTransport();
        } catch (Throwable t) {
            LOG.error(t.getMessage(), t);
            System.exit(1);
        }
    }

    /**
     * Gets client secrets.
     *
     * @return the client secrets
     */
    abstract GoogleClientSecrets getClientSecrets();

    /**
     * Gets credential.
     *
     * @return the credential
     * @throws IOException the io exception
     */
    public abstract Credential getCredential() throws IOException;
}

File: packages/sheets/src/connection/GoogleConnection.ts
/*
 * Licensed to the Software Freedom Conservancy (SFC) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The SFC licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { OAuth2Client, OAuth2ClientOptions } from 'google-auth-library';

/**
 * Represents connection to Google services.
 *
 * @author PhungDucKien
 * @since 2018.04.10
 */
export default abstract class GoogleConnection {
  /**
   * Global instance of the scopes.
   */
  protected static readonly SCOPES = ['https://www.googleapis.com/auth/spreadsheets'];

  /**
   * Gets oAuth2 client options.
   *
   * @return the oAuth2 client options
   */
  public abstract getOAuth2ClientOptions(): Promise<OAuth2ClientOptions>;

  /**
   * Gets auth client.
   *
   * @return the auth client
   */
  public abstract getAuthClient(): Promise<OAuth2Client>;
}
